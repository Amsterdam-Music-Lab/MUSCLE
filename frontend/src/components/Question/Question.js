import React, { useState } from "react";

import ButtonArray from "./_ButtonArray";
import Radios from "./_Radios";
import Range from "./_Range";
import TextRange from "./_TextRange";
import String from "./_String";
import Checkboxes from "./_Checkboxes";
import DropDown from "./_DropDown";
import AutoComplete from "./_AutoComplete";

export const AUTOCOMPLETE = "AUTOCOMPLETE";
export const BUTTON_ARRAY = "BUTTON_ARRAY";
export const CHECKBOXES = "CHECKBOXES";
export const DROPDOWN = "DROPDOWN";
export const RADIOS = "RADIOS";
export const RANGE = "RANGE";
export const TEXT_RANGE = "TEXT_RANGE";
export const STRING = "STRING";

// Question is an experiment view that shows a question and handles storing the answer
const Question = ({
    question,
    onChange,
    id,
    active,
    style,
    emphasizeTitle,
}) => {
    const [value, setValue] = useState(question.value || "");

    const registerChange = (value) => {
        onChange(value, id);
        setValue(value);
    };

    // render view
    const render = (view) => {
        const attrs = {
            value,
            question,
            active,
            style,
            emphasizeTitle,
            onChange: registerChange,
        };

        switch (view) {
            case BUTTON_ARRAY:
                return <ButtonArray {...attrs} />;
            case CHECKBOXES:
                return <Checkboxes {...attrs} />;
            case DROPDOWN:
                return <DropDown {...attrs} />;
            case AUTOCOMPLETE:
                return <AutoComplete {...attrs} />;
            case RADIOS:
                return <Radios {...attrs} />;
            case RANGE:
                return <Range {...attrs} />;
            case TEXT_RANGE:
                return <TextRange {...attrs} />;
            case STRING:
                return <String {...attrs} />;

            default:
                return <div>Unknown question view {view}</div>;
        }
    };

    return (
        <div className="aha__question">
            {question.explainer && (
                <p className="explainer">{question.explainer}</p>
            )}
            <div className="question">{render(question.view)}</div>
        </div>
    );
};

export default Question;
