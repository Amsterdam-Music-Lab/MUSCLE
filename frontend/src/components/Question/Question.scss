@import "./_ButtonArray";
@import "./_Checkboxes";
@import "./_Radios";
@import "./_Range";
@import "./_DropDown";
@import "./_String";

.aha__question {
    min-height: 100%;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-between;
    
    .question {
        flex-grow: 1;
        width: 100%;
    }
}
