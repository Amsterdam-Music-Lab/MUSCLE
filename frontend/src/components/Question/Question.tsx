import React, { useState } from "react";

import classNames from "classnames";

import ButtonArray from "./_ButtonArray";
import Radios from "./_Radios";
import Range from "./_Range";
import TextRange from "./_TextRange";
import IconRange from "./_IconRange";
import String from "./_String";
import Checkboxes from "./_Checkboxes";
import DropDown from "./_DropDown";
import AutoComplete from "./_AutoComplete";

export const AUTOCOMPLETE = "AUTOCOMPLETE";
export const BUTTON_ARRAY = "BUTTON_ARRAY";
export const CHECKBOXES = "CHECKBOXES";
export const DROPDOWN = "DROPDOWN";
export const RADIOS = "RADIOS";
export const RANGE = "RANGE";
export const TEXT_RANGE = "TEXT_RANGE";
export const ICON_RANGE = "ICON_RANGE";
export const STRING = "STRING";

interface Question {
    view: string;
    value: any;
    style: any;
    question: string;
    explainer: string;
    expected_response: string;
}

interface QuestionProps {
    question: Question;
    onChange: (value: any, id: string) => void;
    id: string;
    active: boolean;
    emphasizeTitle?: boolean;
}


// Question is an experiment view that shows a question and handles storing the answer
const Question = ({
    question,
    onChange,
    id,
    disabled = false,
    emphasizeTitle = false,
}: QuestionProps) => {
    const [value, setValue] = useState(question.value || "");

    const registerChange = (value) => {
        onChange(value, id);
        setValue(value);
    };

    // render view
    const render = (view) => {
        const attrs = {
            value,
            question,
            disabled,
            style: question.style,
            emphasizeTitle,
            onChange: registerChange,
        };

        switch (view) {
            case BUTTON_ARRAY:
                return <ButtonArray {...attrs} />;
            case CHECKBOXES:
                return <Checkboxes {...attrs} />;
            case DROPDOWN:
                return <DropDown {...attrs} />;
            case AUTOCOMPLETE:
                return <AutoComplete {...attrs} />;
            case RADIOS:
                return <Radios {...attrs} />;
            case RANGE:
                return <Range {...attrs} />;
            case TEXT_RANGE:
                return <TextRange {...attrs} />;
            case ICON_RANGE:
                return <IconRange {...attrs} />;
            case STRING:
                return <String {...attrs} />;

            default:
                return <div>Unknown question view {view}</div>;
        }
    };

    return (
        <div className="aha__question">
            {question.explainer && (
                <p className="explainer">{question.explainer}</p>
            )}
            <h3 className={classNames({title: emphasizeTitle})}>{question.question}</h3>
            <div className="question">{render(question.view)}</div>
            {question.expected_response &&
                /* Will only be visible when the backend settings has TESTING=True */
                <p className="expected-response">{question.expected_response}</p>
            }
        </div>
    );
};

export default Question;
