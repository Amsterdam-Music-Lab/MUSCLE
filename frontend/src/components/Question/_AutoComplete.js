import React from "react";
import classNames from "classnames";
import Select from "react-select";

// AutoComplete is a question view for selecting a single option from a dropdown list with autocompletion
const AutoComplete = ({
    question,
    value,
    onChange,
    emphasizeTitle = false,
}) => {
    const options = Object.keys(question.choices).map((val, index) => ({
        value: val,
        label: question.choices[val],
    })).sort((a,b)=>(a.label.localeCompare(b.label)));

    return (
        <div className="aha__autocomplete">
            {question.explainer && (
                <p className="explainer">{question.explainer}</p>
            )}
            <h3 className={classNames({ title: emphasizeTitle })}>
                {question.question}
            </h3>
            <div className="control">
                <Select
                    options={options}
                    tabIndex="0"
                    name={question.key}
                    value={options.find(option => option.value === value)}
                    onChange={(choice) => {
                        onChange(choice.value);
                    }}
                />
            </div>
        </div>
    );
};

export default AutoComplete;
