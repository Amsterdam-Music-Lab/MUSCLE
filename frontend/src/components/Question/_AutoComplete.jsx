import React from "react";
import Select from "react-select";

// When using Safari iOS the browser may scroll the page when using the react-select Select.
// Therefore we scroll back to top when the select is blurred
const scrollToTop = () => {
    window.scrollTo(0, 0);
};

// AutoComplete is a question view for selecting a single option from a dropdown list with autocompletion
const AutoComplete = ({ question, value, onChange, emphasizeTitle = false }) => {
    const options = Object.keys(question.choices)
        .map((val, index) => ({
            value: val,
            label: question.choices[val],
        }))
        .sort((a, b) => a.label.localeCompare(b.label));

    return (
        <div className="aha__autocomplete">
            <div className="control">
                <Select
                    options={options}
                    tabIndex="0"
                    name={question.key}
                    value={options.find((option) => option.value === value)}
                    onChange={(choice) => {
                        onChange(choice.value);
                    }}
                    onBlur={scrollToTop}
                />
            </div>
        </div>
    );
};

export default AutoComplete;
