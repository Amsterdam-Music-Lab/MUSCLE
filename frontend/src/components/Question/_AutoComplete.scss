.aha__autocomplete{
  $color: $teal;
  display: flex;
  flex-direction: column;

  .control{
    color: $black;
  }
}
