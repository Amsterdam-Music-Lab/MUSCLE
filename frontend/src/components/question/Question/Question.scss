@import "@/scss/variables";
@import "./_AutoComplete";
@import "./_ButtonArray";
@import "./_Range";
@import "./_DropDown";
@import "./_String";
@import "./_TextRange";

.aha__question {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  margin-bottom: 2rem;

  .question {
    flex-grow: 1;
    width: 100%;
  }

  h3 {
    width: 100%;
    margin-bottom: 1em;

    &.emphasize-title {
      text-align: left;
    }
  }
}

.expected-response {
  display: none;
}
