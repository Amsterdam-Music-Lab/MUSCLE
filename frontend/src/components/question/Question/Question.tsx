/**
 * Copyright (c) 2025 Amsterdam Music Lab
 * SPDX-License-Identifier: MIT
 *
 * This file is part of the MUSCLE project by Amsterdam Music Lab.
 * Licensed under the MIT License. See LICENSE file in the project root.
 */

import type { HTMLAttributes } from "react";
import type IQuestion from "@/types/Question";

import classNames from "classnames";
import QuestionInputFactory from "./QuestionInputFactory";
import styles from "./Question.module.scss";

export interface QuestionProps extends HTMLAttributes<HTMLDivElement> {
  question: IQuestion;

  /**
   * A unique identifier for the question. Will be passed to the
   * onChange callback.
   */
  id: number;

  /**
   * Callback called whenever the input changes. The callback
   * receives both the updatedValue and the id of the question.
   */
  onChange: (updatedValue: string | number | boolean, id: number) => void;

  /** Whether the question is disabled. Default false. */
  disabled?: boolean;
}

/**
 * Question component that shows the question text plus the input component.
 * It uses a factory component that the input based on the view specified in
 * question.view
 */
const Question = ({
  question: questionObj,
  onChange,
  id,
  disabled = false,
  className,
  ...divProps
}: QuestionProps) => {
  // Rename variables internally
  const {
    explainer,
    question,
    style: questionClassName,
    expected_response: expectedResponse,
  } = questionObj;
  return (
    <div
      className={classNames(styles.questionContainer, className)}
      {...divProps}
    >
      {explainer && <p className={styles.explainer}>{explainer}</p>}
      {question && (
        <p className={classNames(styles.questionText, questionClassName)}>
          {question}
        </p>
      )}
      <div className={styles.question}>
        <QuestionInputFactory
          question={questionObj}
          disabled={disabled}
          onChange={(updatedValue: string | number | boolean) =>
            onChange(updatedValue, id)
          }
        />
      </div>
      {expectedResponse && (
        /* Will only be visible when the backend settings has TESTING=True */
        <p className={styles.expectedResponse}>{expectedResponse}</p>
      )}
    </div>
  );
};

export default Question;
