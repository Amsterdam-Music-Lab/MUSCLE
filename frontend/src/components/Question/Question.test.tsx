import { render, fireEvent, screen } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import Question, {QuestionProps} from './Question';
import { QuestionViews } from '@/types/Question';

describe('Question Component', () => {
    const mockOnChange = vi.fn();

    const defaultProps = {
        question: {
            key: 'test-question',
            question: 'Test Question',
            view: QuestionViews.STRING,
            value: '',
            choices: {
                '1': 'One',
                '2': 'Two',
            },
            style: {}
        },
        onChange: mockOnChange,
        id: 42,
    };

    const getProps = (props = {}) => ({ ...defaultProps, ...props });

    it('renders the question text', () => {
        render(<Question {...defaultProps} />);
        expect(screen.getByText('Test Question')).toBeTruthy();
    });

    it('renders an explainer when provided', () => {
        const props = {
            ...defaultProps,
            question: {
                ...defaultProps.question,
                explainer: 'This is an explainer',
            },
        };
        render(<Question {...props} />);
        expect(screen.getByText('This is an explainer')).toBeTruthy();
    });

    it('renders the expected response when provided and TESTING is true', () => {
        const props = {
            ...defaultProps,
            question: {
                ...defaultProps.question,
                expected_response: 'Expected response',
            },
        };
        // Mock the TESTING environment variable
        vi.stubGlobal('process', { env: { TESTING: 'true' } });
        render(<Question {...props} />);
        expect(screen.getByText('Expected response')).toBeTruthy();
        vi.unstubAllGlobals();
    });

    it('calls onChange when the value changes', () => {
        render(<Question {...defaultProps} />);
        const input = screen.getByRole('textbox');
        fireEvent.change(input, { target: { value: 'New Value' } });
        expect(mockOnChange).toHaveBeenCalledWith('New Value', 42);
    });

    it('applies classNames if question.style is defined', () => {
        const props = defaultProps;
        props.question.style = 'some-style';
        const { container } = render(<Question {...props} />);
        expect(container.querySelector('.some-style')).toBeTruthy();
    });

    it('disables the input when disabled prop is true', () => {
        const props = getProps({
            disabled: true,
            question: {
                ...defaultProps.question,
                view: QuestionViews.BUTTON_ARRAY,
                choices: { '1': 'One', '2': 'Two' },
            }
        });
        render(<Question {...props} />);
        const input = screen.getByTestId('toggle-button-1');
        expect(input).toBeTruthy();
        expect(input.attributes.getNamedItem('disabled')).toBeTruthy();
    });
});
