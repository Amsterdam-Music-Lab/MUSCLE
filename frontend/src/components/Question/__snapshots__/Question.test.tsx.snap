// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Question Component > renders different question views based on the view prop 1`] = `
<div
  class="aha__question"
>
  <h3
    class=""
  >
    Test Question
  </h3>
  <div
    class="question"
  >
    <div
      class="aha__buttons buttons d-flex flex-wrap justify-content-center p-3 w-100"
    >
      <div
        aria-label="Button Array"
        class="btn-group-toggle-custom"
        data-toggle="buttons"
        role="group"
      >
        <label
          class="btn btn-secondary btn-lg"
          tabindex="0"
        >
          <input
            aria-checked="false"
            class="1"
            data-testid="toggle-button-1"
            id="0"
            type="radio"
            value="1"
          />
          One
        </label>
        <label
          class="btn btn-secondary btn-lg"
          tabindex="0"
        >
          <input
            aria-checked="false"
            class="2"
            data-testid="toggle-button-2"
            id="1"
            type="radio"
            value="2"
          />
          Two
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`Question Component > renders different question views based on the view prop 2`] = `
<div
  class="aha__question"
>
  <h3
    class=""
  >
    Test Question
  </h3>
  <div
    class="question"
  >
    <div
      class="aha__checkboxes"
    >
      <div
        class="checkbox"
        tabindex="0"
      >
        <i />
        <span>
          One
        </span>
      </div>
      <div
        class="checkbox"
        tabindex="0"
      >
        <i />
        <span>
          Two
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Question Component > renders different question views based on the view prop 3`] = `
<div
  class="aha__question"
>
  <h3
    class=""
  >
    Test Question
  </h3>
  <div
    class="question"
  >
    <div
      class="aha__dropdown"
    >
      <select
        tabindex="0"
      >
        <option
          value=""
        />
        <option
          value="1"
        >
          One
        </option>
        <option
          value="2"
        >
          Two
        </option>
      </select>
    </div>
  </div>
</div>
`;

exports[`Question Component > renders different question views based on the view prop 4`] = `
<div
  class="aha__question"
>
  <h3
    class=""
  >
    Test Question
  </h3>
  <div
    class="question"
  >
    <div
      class="aha__autocomplete"
    >
      <div
        class="control"
      >
        <div
          class=" css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-2-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class=" css-13cymwt-control"
          >
            <div
              class=" css-1fdsijx-ValueContainer"
            >
              <div
                class=" css-1jqq78o-placeholder"
                id="react-select-2-placeholder"
              >
                Select...
              </div>
              <div
                class=" css-qbdosj-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-2-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-2-input"
                  role="combobox"
                  spellcheck="false"
                  style="color: inherit; opacity: 1; width: 100%; grid-area: 1 / 2; font: inherit; min-width: 2px; border: 0px; margin: 0px; outline-color: initial; outline-style: initial; outline-width: 0px; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class=" css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Question Component > renders different question views based on the view prop 5`] = `
<div
  class="aha__question"
>
  <h3
    class=""
  >
    Test Question
  </h3>
  <div
    class="question"
  >
    <div
      class="aha__radios"
    >
      <div
        class="radio"
        tabindex="0"
      >
        <i />
        <span>
          One
        </span>
      </div>
      <div
        class="radio"
        tabindex="0"
      >
        <i />
        <span>
          Two
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Question Component > renders different question views based on the view prop 6`] = `
<div
  class="aha__question"
>
  <h3
    class=""
  >
    Test Question
  </h3>
  <div
    class="question"
  >
    <div
      class="aha__range empty"
    >
      <h1
        class="current-value"
      >
        â†”
      </h1>
      <div
        aria-orientation="horizontal"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="NaN"
        class="rangeslider rangeslider-horizontal"
      >
        <div
          class="rangeslider__fill"
          style=""
        />
        <div
          class="rangeslider__handle"
          style=""
          tabindex="0"
        >
          <div
            class="rangeslider__handle-label"
          />
        </div>
        <ul
          class="rangeslider__labels"
        />
      </div>
      <div
        class="limits"
      >
        <span
          class="min"
        />
        <span
          class="max"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Question Component > renders different question views based on the view prop 7`] = `
<div
  class="aha__question"
>
  <h3
    class=""
  >
    Test Question
  </h3>
  <div
    class="question"
  >
    <div
      class="aha__text-range empty"
    >
      <div>
        <h4
          class="current-value"
          style="position: relative; left: 0%;"
        >
          <span
            class="fa-solid fa-arrows-left-right fa-2x"
          />
        </h4>
      </div>
      <div
        aria-orientation="horizontal"
        aria-valuemax="10"
        aria-valuemin="0"
        aria-valuenow="5"
        class="rangeslider rangeslider-horizontal"
      >
        <div
          class="rangeslider__fill"
          style="width: 0px;"
        />
        <div
          class="rangeslider__handle"
          style="left: 0px;"
          tabindex="0"
        >
          <div
            class="rangeslider__handle-label"
          />
        </div>
        <ul
          class="rangeslider__labels"
        />
      </div>
      <div
        class="limits"
      >
        <div>
          <span
            class="min"
          >
            One
          </span>
          <span
            class="max"
          >
            Two
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Question Component > renders different question views based on the view prop 8`] = `
<div
  class="aha__question"
>
  <h3
    class=""
  >
    Test Question
  </h3>
  <div
    class="question"
  >
    <div
      class="aha__text-range empty"
    >
      <div>
        <h4
          class="current-value"
          style="position: relative; left: -44%;"
        >
          <span
            class="fa-solid fa-arrows-left-right fa-2x"
          />
        </h4>
      </div>
      <div
        aria-orientation="horizontal"
        aria-valuemax="1"
        aria-valuemin="0"
        aria-valuenow="0"
        class="rangeslider rangeslider-horizontal"
      >
        <div
          class="rangeslider__fill"
          style="width: 0px;"
        />
        <div
          class="rangeslider__handle"
          style="left: 0px;"
          tabindex="0"
        >
          <div
            class="rangeslider__handle-label"
          />
        </div>
        <ul
          class="rangeslider__labels"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Question Component > renders different question views based on the view prop 9`] = `
<div
  class="aha__question"
>
  <h3
    class=""
  >
    Test Question
  </h3>
  <div
    class="question"
  >
    <div
      class="aha__string"
    >
      <input
        value=""
      />
    </div>
  </div>
</div>
`;
